### PR TITLE
Enhancement/pattern expression

### PR DESCRIPTION
**Pull request by czpmango**
## What type of PR is this?
- [ ] bug
- [ ] feature
- [x] enhancement

#### Description:
This pr is optimized for the case of pattern expression as a predicate. In this case, the runtime doesn't actually have to construct the path. The principle is similar to sql semi apply related optimization. 

Compared to previous implementations, this pr has three main optimization points:
- The runtime does not construct paths to pattern predicate(overhead of Project(pathBuild expression)).
- Avoid the overhead of RollUpJoin, which is likely to have high cpu overhead(hash build and probe)and the large table being constructed will also have high memory overhead.
- Avoid the overhead of Filter executor.

For example:
Query statement: `MATCH (v:player)-->(n) where (v)-[:like]->() and size((v)-[:serve]-())>2 and v.age>3 and not (v)-[:teammate]-(n) RETURN id(v)`
Old plan: 
graphviz (3)
Optimized plan:
graphviz (4)

## Special notes for your reviewer
I will fix some FIXMEs related to handling of expression later. Expression visitor needs to cover all expressions, and currently supports only simple expressions.

## Checklist:
Tests:
- [ ] Unit test(positive and negative cases)
- [ ] Function test
- [x] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:
Optimize the implementation of pattern predicate.

### CODE REVIEW

**Review by jievince — APPROVED** [review]
Well done!

**Comment by czpmango** [comment]
We should add PushFilterDownRollUpApply and PushFilterDownPatternApply rules.

**Comment by czpmango** [comment]
> Please add test.

This pr can be considered as an optimization or refactoring, and all the statement behavior is guaranteed by previous tests related to pattern expression. 
If the test case coverage is not complete, it will be evaluated and completed in the future planning.

**Review by jievince — APPROVED** [review]

File: src/graph/planner/match/MatchPathPlanner.cpp [hunk 1]
```diff
@@ -97,7 +97,11 @@ StatusOr<SubPlan> MatchPathPlanner::transform(
                             subplan,
                             nodeAliasesSeenInPattern));
 
-  MatchSolver::buildProjectColumns(qctx, path, subplan);
+  // No need to actually build path if the path is just a predicate
+  if (!path.isPred) {
```
Comment (Shylock-Hg) [inline src/graph/planner/match/MatchPathPlanner.cpp:101]: Better to do it in optimize rule.

Comment (czpmango) [reply src/graph/planner/match/MatchPathPlanner.cpp:101]: Generally ACK.
But IMHO, pattern expression and pattern predicate are semantically different, so it is also reasonable to generate different execution plans in the planner.

Comment (Shylock-Hg) [reply src/graph/planner/match/MatchPathPlanner.cpp:101]: Not, in general, it's properties prune problem. e.g. `where (v)-[]->()` means `List<Path>` is not empty.

Comment (Shylock-Hg) [reply src/graph/planner/match/MatchPathPlanner.cpp:101]: Planner should generate the most generic plan, and keep simple. Too many branches are evil.

Comment (czpmango) [reply src/graph/planner/match/MatchPathPlanner.cpp:101]: This pr does trigger some optimization related to property pruning, but does it have anything to do with pattern predicate. I don't get you about "in general, it's properties prune problem".

Comment (czpmango) [reply src/graph/planner/match/MatchPathPlanner.cpp:101]: > Planner should generate the most generic plan, and keep simple. Too many branches are evil.

Just one more if-block. Is it necessary to implement it in the optimizer phase and add a few hundred lines of code?

Comment (czpmango) [reply src/graph/planner/match/MatchPathPlanner.cpp:101]: I also approve of doing that in optimizer, but I still stand by my opinion.
It is true that planner should generate the most general plan, but it does not mean that a designed operator should never be generated in planner. For example, if the pattern(like (v)-[e]-(v)) is already a ExpandInto pattern, is it necessary to convert ExpandAll into ExpandInto in optimizer phase when all information is already obvious in the AST.

**Review by jievince — APPROVED** [review]

File: src/graph/planner/match/MatchPathPlanner.cpp (hunk 1 above)
Comment (Shylock-Hg) [reply src/graph/planner/match/MatchPathPlanner.cpp:101]: ok

**Review by Shylock-Hg — APPROVED** [review]

**Comment by yixinglu** [comment]
Now we don't hurry to tuning the LDBC according to the PR, you could transform the implementation to the optimizer rule.

**Comment by czpmango** [comment]
> Now we don't hurry to tuning the LDBC according to the PR, you could transform the implementation to the optimizer rule.

IMHO, implementing this in the optimizer is not necessarily better. 
https://github.com/vesoft-inc/nebula/pull/4916#discussion_r1032056777

**Comment by Sophie-Xie** [comment]
Pls don‘t merge, wait for @yixinglu .

**Review by yixinglu — DISMISSED** [review]

**Review by jievince — DISMISSED** [review]

**Review by jievince — APPROVED** [review]

**Review by yixinglu — COMMENTED** [review]
Need some tck cases

**Comment by czpmango** [comment]
> Need some tck cases

As it says here: https://github.com/vesoft-inc/nebula/pull/4916#issuecomment-1325870778

**Review by yixinglu — APPROVED** [review]

**Review by jievince — APPROVED** [review]

**Pull request merged**